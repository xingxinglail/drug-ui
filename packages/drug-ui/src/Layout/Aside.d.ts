import * as React from 'react';
import { LayoutProps } from './Layout';
export interface AsideProps extends LayoutProps {
    width?: number | string;
}
declare const Aside: React.FC<AsideProps>;
export default Aside;
