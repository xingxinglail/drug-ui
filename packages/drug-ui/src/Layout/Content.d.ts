import * as React from 'react';
import { LayoutProps } from './Layout';
declare const Content: React.FC<LayoutProps>;
export default Content;
