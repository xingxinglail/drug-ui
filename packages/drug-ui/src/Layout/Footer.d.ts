import * as React from 'react';
import { LayoutProps } from './Layout';
declare const Footer: React.FC<LayoutProps>;
export default Footer;
