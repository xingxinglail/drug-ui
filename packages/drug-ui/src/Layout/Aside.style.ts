import { Style, Styles } from 'jss';
import { Theme } from '../styles';

const root: Style = {
    position: 'relative',
    backgroundColor: '#001529'
};

export const styles = (_: Theme): Styles => ({
    root
});
