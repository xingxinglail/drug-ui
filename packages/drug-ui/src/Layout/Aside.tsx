import * as React from 'react';
import * as PropTypes from 'prop-types';
import classnames from 'classnames';
import { styles } from './Aside.style';
import { createUseStyles } from '../styles';
import { LayoutProps } from './Layout';
import { isNumeric } from '../utils';

export interface AsideProps extends LayoutProps {
    width?: number | string;
}

type LayoutClassProps = 'root';

const name = 'LayoutAside';

const useStyles = createUseStyles<LayoutClassProps>(styles, name);

const Aside: React.FC<AsideProps> = React.forwardRef<HTMLDivElement, AsideProps>((props, ref) => {
    const { className, width, children, ...rest } = props;
    const value = isNumeric(width) ? `${ width }px` : width;
    const widthStyle = {
        width: value,
        flex: `0 0 ${ value }`
    };

    const classes = useStyles();
    const classNames = classnames(
        classes.root,
        className
    );

    return (
        <aside
            className={ classNames }
            style={ widthStyle }
            ref={ ref }
            { ...rest }>
            { children }
        </aside>
    );
});

Aside.displayName = name;

Aside.defaultProps = {
    className: undefined,
    width: 200
};

Aside.propTypes = {
    className: PropTypes.string,
    width: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
    ])
};

export default Aside;
