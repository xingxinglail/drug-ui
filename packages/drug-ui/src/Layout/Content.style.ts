import { Style, Styles } from 'jss';
import { Theme } from '../styles';

const root: Style = {
    flex: 'auto'
};

export const styles = (_: Theme): Styles => ({
    root
});
