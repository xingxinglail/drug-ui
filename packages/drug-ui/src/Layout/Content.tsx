import * as React from 'react';
import classnames from 'classnames';
import { styles } from './Content.style';
import { createUseStyles } from '../styles';
import { LayoutProps } from './Layout';

type LayoutClassProps = 'root';

const name = 'LayoutContent';

const useStyles = createUseStyles<LayoutClassProps>(styles, name);

const Content: React.FC<LayoutProps> = React.forwardRef<HTMLDivElement, LayoutProps>((props, ref) => {
    const { className, children, ...rest } = props;
    const classes = useStyles();

    const classNames = classnames(
        classes.root,
        className
    );

    return (
        <main
            className={ classNames }
            ref={ ref }
            { ...rest }>
            { children }
        </main>
    );
});

Content.displayName = name;

export default Content;
