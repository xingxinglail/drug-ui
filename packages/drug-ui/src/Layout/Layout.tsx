import * as React from 'react';
import { Styles } from 'jss';
import classnames from 'classnames';
import { styles } from './Layout.style';
import { createUseStyles } from '../styles';
import Header from './Header';
import Content from './Content';
import Aside, { AsideProps } from './Aside';
import Footer from './Footer';

export interface LayoutProps extends React.HTMLAttributes<HTMLDivElement> {
    ref?: React.Ref<HTMLDivElement>;
}

type LayoutClassProps = 'root' | 'hasAside';

const name = 'Layout';

const useStyles = createUseStyles<LayoutClassProps>(styles, name);

interface IProps {
    children: (classes: Styles) => React.ReactElement;
}

const Classes = ({ children }: IProps) => {
    const classes: Styles = useStyles();
    return children(classes);
};

class Basic extends React.Component<LayoutProps, any> {

    static displayName = name;

    static Header: any;

    static Footer: any;

    static Content: any;

    static Aside: any;

    render () {
        const { className, children, ...rest } = this.props;
        let hasAside = false;
        if (children !== undefined && children !== null) {
            let _children: React.ReactElement[] = [];
            Array.isArray(children) ? _children = (children as React.ReactElement[]) : _children.push(children as React.ReactElement);
            hasAside = (_children as React.ReactElement[]).some(node => {
                return node.type === Aside;
            });
        }

        return (
            <Classes>
                { (classes: Styles) => (
                    <section
                        className={ classnames(classes.root, { [classes.hasAside]: hasAside }, className) }
                        { ...rest }>
                        { children }
                    </section>
                ) }
            </Classes>
        );
    }
}

const Layout: React.ComponentClass<LayoutProps> & {
    Header: React.FC<LayoutProps>;
    Content: React.FC<LayoutProps>;
    Aside: React.FC<AsideProps>;
    Footer: React.FC<LayoutProps>;
} = Basic;

Layout.Header = Header;
Layout.Content = Content;
Layout.Aside = Aside;
Layout.Footer = Footer;

export default Layout;
