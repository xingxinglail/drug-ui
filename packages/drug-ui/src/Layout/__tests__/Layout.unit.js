import * as renderer from 'react-test-renderer';
import * as React from 'react';
import * as enzyme from 'enzyme';
import Layout from '..';

const mount = enzyme.mount;

const { Header, Content, Aside, Footer } = Layout;

describe('<Layout />', () => {

    it('render successful', () => {
        const json = renderer.create(
            <Layout>
                <Header>Header</Header>
                <Layout>
                    <Aside>Aside</Aside>
                    <Content>Content</Content>
                </Layout>
                <Footer>Footer</Footer>
            </Layout>
        ).toJSON();
        expect(json).toMatchSnapshot();
    });

    it('Layout 设置 className', () => {
        const wrapper = mount(
            <Layout className="layout">layout</Layout>
        );
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains('layout')).toBeTruthy();
    });

    it('Header 设置 className', () => {
        const wrapper = mount(
            <Header className="header">Header</Header>
        );
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains('header')).toBeTruthy();
    });

    it('Aside 设置 className', () => {
        const wrapper = mount(
            <Aside className="aside">Aside</Aside>
        );
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains('aside')).toBeTruthy();
    });

    it('Content 设置 className', () => {
        const wrapper = mount(
            <Content className="content">Content</Content>
        );
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains('content')).toBeTruthy();
    });

    it('Footer 设置 className', () => {
        const wrapper = mount(
            <Footer className="footer">Footer</Footer>
        );
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains('footer')).toBeTruthy();
    });

    it('Aside 设置 width', () => {
        const wrapper = mount(
            <Aside width={ 300 }>Aside</Aside>
        );
        const dom = wrapper.getDOMNode();
        expect(getComputedStyle(dom).width).toBe('300px');
        expect(getComputedStyle(dom).flex).toBe('0 0 300px');
    });
});
