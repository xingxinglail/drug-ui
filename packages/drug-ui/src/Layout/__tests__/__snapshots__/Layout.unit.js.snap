// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> render successful 1`] = `
<section
  className="Layout-root-0-2-1"
>
  <header
    className="LayoutHeader-root-0-2-3"
  >
    Header
  </header>
  <section
    className="Layout-root-0-2-1 Layout-hasAside-0-2-2"
  >
    <aside
      className="LayoutAside-root-0-2-4"
      style={
        Object {
          "flex": "0 0 200px",
          "width": "200px",
        }
      }
    >
      Aside
    </aside>
    <main
      className="LayoutContent-root-0-2-5"
    >
      Content
    </main>
  </section>
  <footer
    className="LayoutFooter-root-0-2-6"
  >
    Footer
  </footer>
</section>
`;
